File: algorithms/easy/migratoryBirds.py
# Problem definition: https://www.hackerrank.com/challenges/migratory-birds/problem

def initializeMemory(idBirds: list) -> dict:
  memory = {}
  for i in range(min(idBirds), max(idBirds) + 1):
    memory[str(i)] = 0
  return memory


def migratoryBirds(idBirds: list) -> int:
  max = idBirds[0]
  memory = initializeMemory(idBirds)
  for i in idBirds:
    memory[str(i)] = memory[str(i)] + 1
    if memory[str(i)] > memory[str(max)]:
      max = i
    elif memory[str(i)] == memory[str(max)]:
      if max > i:
        max = i

  return max


print(migratoryBirds([1, 1, 2, 2, 3])) # Expected output: 1   
print(migratoryBirds([1, 4, 4, 5, 5, 3])) # Expected output: 4
print(migratoryBirds([1, 2, 3, 4, 5, 4, 3, 2, 1, 3, 4])) # Expected output: 3

File: algorithms/easy/migratoryBirds.cpp
// Problem reference: https://www.hackerrank.com/challenges/migratory-birds/problem
// Predefined complexities: size = O(1), vector.begin = O(1), vector.erase(0) = O(1)
// Code complexity: ...
#include <iostream>
#include <bits/stdc++.h>

using intMap = std::unordered_map<int, int>;
using intVec = std::vector<int>;


intMap initializeMemory(intVec birds){
  intMap map;
  int maxVec = *max_element(std::begin(birds), std::end(birds));
  int minVec = *min_element(std::begin(birds), std::end(birds));

  for(int i = minVec; i <= maxVec; i++){
    map[i] = 0;
  }

  return map;
}


int migratoryBirds(intVec birds){

}

int main(){
  intMap map = initializeMemory({1,2,3});
  for(auto item : map){
    std::cout << item.first << " " << item.second << "\n";
  }
}